### PR TITLE
feat(button): add tooltip prop (#255) (#258)

### LINKED ISSUE #255: A good way to provide more info to users on disabled actions
This issue is wrapper issue that consists of many different tasks so that we can solve a particular problem.

## What we wanna do

Many apps in GB has quite complex authorizations to perform some actions like updating objects. When user is not permitted to perform an action, we tend to make a button disabled like this.

Screenshot 2023-04-20 at 9 45 06

The above example shows that "Update tags" button is not available.

The issue is, the reason why it's disabled is not clear to users. Sometimes, the button is disabled because the user is lacking the authorization permission. But sometimes, maybe the button is disabled because the object state is not right (e.g. If an object has status "Completed", then user should not be able to update it).

Simply disabling or hiding button make users wonder why? And we would like to have some way to nicely handle this situation.

## Clickable tooltip

One thing that comes to our mind very quickly is to use tooltips. However, showing tooltips on hover is not a good idea in terms of touch devices. It's almost impossible to show tooltips on button.

So, instead of showing the tooltip on hover, how about if we show tooltip when user actually click the disabled button?

Screenshot 2023-04-20 at 10 28 22

This way we can add description to actions that needs explanation to understand why something is permitted or not.

Sidenote: We should also consider doing the same for `:info` option on input components (make it clickable tooltip rather than activate on hover) 👉 https://github.com/globalbrain/sefirot/issues/259

## Todos

Related issues in order to make this work.

- [x] https://github.com/globalbrain/sefirot/issues/256
- [x] https://github.com/globalbrain/sefirot/issues/257
- [x] https://github.com/globalbrain/sefirot/issues/258

### LINKED ISSUE #258: [Button] Add option to show tooltip when clicking disabled button
- ref: #255
- require: https://github.com/globalbrain/sefirot/issues/256
- require: https://github.com/globalbrain/sefirot/issues/257

In order to easily achieve #255, it would be nice to have easy option to do this. Maybe something like:

```ts
interface Props {
  // When the message is set, show tooltip on `@disabled-click`.
  disabledMsg?: string
}
```

```html
<SButton
  label="Delete"
  disabledMsd="You are not authorized."
/>

<!-- Also via #disabled-msg slot -->
<SButton label="Delete">
  <template #disabled-msg>
    You are not authorized.
  </template>
</SButton>
```

### PR DESCRIPTION
**Pull request by brc-dd**
fixes #258
closes #255

### CODE REVIEW

**Review by kiaking — COMMENTED** [review]
Looking nice! I like the API. We might wanna have easier way to define a tooltip message in the future, because currently we have to write few lines of codes to switch the tooltip message depending on button state, but... It's not clear right now.

So I think it would be safer to start primitive, and add something more when we have better idea of how API should look like.

---

Would it be possible to adjust the tooltip position? Currently, it's over wrapping with button.

Screenshot 2023-04-26 at 18 01 02

I want tooltip to have maybe 4px margin between button. If this is something needs to be done in `STooltip` side, we can work on it on another PR 👀

File: lib/components/Fragment.vue
[the remark this answers is not in the document]
Comment (brc-dd) [reply lib/components/Fragment.vue:1]: Eduardo wrote it 😂 It was there in one of the Vue RFCs.

**Comment by brc-dd** [comment]
> We might wanna have easier way to define a tooltip message in the future.

Instead of passing as object we probably can flatten that? `tooltipText="..."`, `tooltipPosition="..."`. WDYT?

**Comment by brc-dd** [comment]
> I want tooltip to have maybe 4px margin between button.

This already is there, but in that story the button is near the top so our positioning algorithm is showing it like that. Try on the icons one, there is sufficient space, it should work fine:

image

**Comment by kiaking** [comment]
> > I want tooltip to have maybe 4px margin between button.
> 
> This already is there, but in that story the button is near the top so our positioning algorithm is showing it like that. Try on the icons one, there is sufficient space, it should work fine:

Ahhh it's using absolute position. Gotcha. Thanks! I thought it would go up to "Vie documentation" section 😅

**Comment by kiaking** [comment]
> > We might wanna have easier way to define a tooltip message in the future.
> 
> Instead of passing as object we probably can flatten that? `tooltipText="..."`, `tooltipPosition="..."`. WDYT?

Hmmmmm tough choice. I think current object might be better... not 100% sure but because I think we'll be probably doing these kind of stuff.

```ts
const disabled = ref(false)

const buttonTooltip = computedWhen(disabled, () => ({
  text: 'You are not authorized',
  position: 'top'
}))
```

And maybe in the future, we might wanna add more option to tooltip (like to include info icon and such maybe?), in that case we don't want to keep adding props I guess 🤔

File: lib/components/Fragment.vue
[the remark this answers is not in the document]
Comment (kiaking) [reply lib/components/Fragment.vue:1]: OK hooray to Posva 🫡

**Comment by brc-dd** [comment]
How about making `tooltip.text` a `MaybeRef`? That way you won't need computing everything, just text being a ref/computed would work. If it's empty (and slot is not there) then we don't show the tooltip.

**Comment by kiaking** [comment]
> And how about making tooltip.text a MaybeRef? That way you won't need computing everything, just text being a ref/computed would work. If it's falsy (and slot is not there) then we don't show the tooltip.

OK that's genius. Very nice idea. Let's do this 👍

**Comment by kiaking** [comment]
@brc-dd
Pushed several things! Could you take a look 🙏 

1. Added `:display` prop to `<STooltip>` in order to make `.STooltip` element `display: inline-block`.
    - When we wrap default `<SButton>` which is `display: inline-flex`, with `display: inline` which is default for `<STooltip>` as `<span>`, the height will not be fully calculated. So I'm setting `<STooltip>` display as `inline-block` here.
3. Added more control to `<SButton>` story.
4. Added/Updated docs for `<SButton>`, `<STooltip>`, and `<SFragment>`.

**Comment by brc-dd** [comment]
Look good to me 🚀

**Comment by kiaking** [comment]
Thanks!

**Pull request merged**